File: packages/client/src/mdp-client.ts
import {
  type AuthContext,
  type CallClientMessage,
  type HttpMethod,
  type PathDescriptor,
  type ServerToClientMessage,
  createSerializedError
} from '@modeldriveprotocol/protocol'

import { createDefaultTransport, resolveServerUrl } from './transport/client-connection.js'
import { ProcedureRegistry } from './runtime/procedure-registry.js'
import {
  createLegacyPromptPath,
  createLegacyResourcePath,
  createLegacySkillPath,
  createLegacyToolPath
} from './runtime/legacy-capability-paths.js'
import { MdpClientReconnectController } from './runtime/reconnect-controller.js'
import { authenticateTransport } from './transport/transport-auth.js'
import type {
  ClientDescriptorOverride,
  ClientInfo,
  ClientTransport,
  ClientTransportAuthOptions,
  ExposeLegacyPromptOptions,
  ExposePathOptions,
  ExposeResourceOptions,
  ExposeToolOptions,
  ExposeLegacySkillOptions,
  LegacyPromptHandler,
  LegacyResourceHandler,
  LegacySkillHandler,
  LegacyToolHandler,
  LegacyCapabilityContext,
  MdpClientOptions,
  PathHandler,
  PathInvocationContext,
  PathRequest,
  PathInvocationMiddleware,
  StaticPathDefinition
} from './types.js'

export class MdpClient {
  private readonly serverUrl: string
  private readonly serverProtocol: string
  private readonly usesDefaultTransport: boolean
  private clientInfo: ClientInfo
  private readonly registry = new ProcedureRegistry()
  private readonly transport: ClientTransport
  private readonly transportAuth: ClientTransportAuthOptions | undefined
  private readonly reconnectController: MdpClientReconnectController
  private auth: AuthContext | undefined
  private connected = false
  private registered = false
  private restoreRegistrationOnReconnect = false

  constructor(options: MdpClientOptions) {
    this.serverUrl = options.serverUrl
    this.serverProtocol = new URL(options.serverUrl).protocol
    this.usesDefaultTransport = options.transport === undefined
    this.clientInfo = options.client
    this.auth = options.auth
    this.transportAuth = options.transportAuth
    this.transport = options.transport ?? createDefaultTransport(options.serverUrl)
    this.reconnectController = new MdpClientReconnectController({
      serverUrl: this.serverUrl,
      reconnect: options.reconnect,
      reconnectTransport: async () => {
        await this.openTransport()

        if (this.reconnectController.isDisconnectRequested()) {
          await this.transport.close()
          return
        }

        if (this.restoreRegistrationOnReconnect) {
          this.sendRegisterMessage()
          this.registered = true
        }
      }
    })
    this.transport.onMessage((message) => {
      void this.handleMessage(message)
    })
    this.transport.onClose(() => {
      void this.handleTransportClose()
    })
  }

  useInvocationMiddleware(middleware: PathInvocationMiddleware): this {
    this.registry.useInvocationMiddleware(middleware)
    return this
  }

  removeInvocationMiddleware(middleware: PathInvocationMiddleware): this {
    this.registry.removeInvocationMiddleware(middleware)
    return this
  }

  expose(
    path: string,
    definition: StaticPathDefinition | ExposePathOptions,
    handler?: PathHandler
  ): this {
    this.registry.expose(path, definition, handler)
    return this
  }

  exposeTool(
    name: string,
    handler: LegacyToolHandler,
    options: ExposeToolOptions = {}
  ): this {
    return this.expose(
      createLegacyToolPath(name),
      {
        method: 'POST',
        ...(options.description ? { description: options.description } : {}),
        ...(options.inputSchema ? { inputSchema: options.inputSchema } : {}),
        ...(options.outputSchema ? { outputSchema: options.outputSchema } : {}),
        ...(options.contentType ? { contentType: options.contentType } : {}),
        legacy: {
          kind: 'tool',
          name
        }
      },
      async (request, context) => handler(readLegacyArgs(request), toLegacyContext(request, context))
    )
  }

  exposePrompt(
    name: string,
    definition: string | LegacyPromptHandler,
    options: ExposeLegacyPromptOptions = {}
  ): this {
    const path = createLegacyPromptPath(name)

    if (typeof definition === 'string') {
      return this.expose(path, {
        ...(options.description ? { description: options.description } : {}),
        ...(options.inputSchema ? { inputSchema: options.inputSchema } : {}),
        ...(options.outputSchema ? { outputSchema: options.outputSchema } : {}),
        legacy: {
          kind: 'prompt',
          name
        }
      }, async () => ({
        messages: [
          {
            role: 'user',
            content: definition
          }
        ]
      }))
    }

    return this.expose(
      path,
      {
        ...(options.description ? { description: options.description } : {}),
        ...(options.inputSchema ? { inputSchema: options.inputSchema } : {}),
        ...(options.outputSchema ? { outputSchema: options.outputSchema } : {}),
        legacy: {
          kind: 'prompt',
          name
        }
      },
      async (request, context) => definition(
        readLegacyArgs(request),
        toLegacyContext(request, context)
      )
    )
  }

  exposeSkill(
    name: string,
    definition: string | LegacySkillHandler | (() => unknown | Promise<unknown>),
    options: ExposeLegacySkillOptions = {}
  ): this {
    const path = createLegacySkillPath(name)

    if (typeof definition === 'string') {
      return this.expose(path, {
        ...(options.description ? { description: options.description } : {}),
        ...(options.contentType ? { contentType: options.contentType } : {}),
        legacy: {
          kind: 'skill',
          name
        }
      }, async () => definition)
    }

    return this.expose(
      path,
      {
        ...(options.description ? { description: options.description } : {}),
        ...(options.contentType ? { contentType: options.contentType } : {}),
        legacy: {
          kind: 'skill',
          name
        }
      },
      async (request, context) => {
        const legacyContext = toLegacyContext(request, context)

        return definition.length <= 0
          ? (definition as () => unknown | Promise<unknown>)()
          : (definition as LegacySkillHandler)(
              request.queries,
              request.headers,
              legacyContext
            )
      }
    )
  }

  exposeResource(
    uri: string,
    handler: LegacyResourceHandler,
    options: ExposeResourceOptions = {}
  ): this {
    return this.expose(
      createLegacyResourcePath(uri),
      {
        method: 'GET',
        ...(options.description ? { description: options.description } : {}),
        ...(options.mimeType ? { contentType: options.mimeType } : {}),
        legacy: {
          kind: 'resource',
          uri,
          ...(options.name ? { name: options.name } : {})
        }
      },
      async (request, context) => handler(toLegacyContext(request, context))
    )
  }

  unexpose(path: string, method?: HttpMethod): this {
    this.registry.unexpose(path, method)
    return this
  }

  async connect(): Promise<void> {
    this.reconnectController.beginConnect()
    await this.reconnectController.connect()
  }

  setAuth(auth?: AuthContext): this {
    this.auth = auth
    return this
  }

  async authenticateTransport(auth: AuthContext | undefined = this.auth): Promise<void> {
    await authenticateTransport({
      serverUrl: this.serverUrl,
      serverProtocol: this.serverProtocol,
      usesDefaultTransport: this.usesDefaultTransport,
      transportAuth: this.transportAuth,
      auth
    })
  }

  register(overrides: ClientDescriptorOverride = {}): void {
    this.ensureConnected()
    this.clientInfo = {
      ...this.clientInfo,
      ...overrides
    }

    this.sendRegisterMessage()
    this.registered = true
  }

  async disconnect(): Promise<void> {
    this.restoreRegistrationOnReconnect = false
    this.reconnectController.beginDisconnect()

    if (this.connected && this.registered) {
      this.transport.send({
        type: 'unregisterClient',
        clientId: this.clientInfo.id
      })
    }

    await this.transport.close()
    this.connected = false
    this.registered = false
  }

  describe() {
    return this.registry.describe(this.clientInfo)
  }

  syncCatalog(paths: PathDescriptor[] = this.registry.describePaths()): void {
    this.ensureRegistered()

    this.transport.send({
      type: 'updateClientCatalog',
      clientId: this.clientInfo.id,
      paths
    })
  }

  private async handleMessage(message: ServerToClientMessage): Promise<void> {
    switch (message.type) {
      case 'ping':
        this.transport.send({
          type: 'pong',
          timestamp: message.timestamp
        })
        return
      case 'callClient':
        await this.handleInvocation(message)
        return
      default:
        return
    }
  }

  private async handleInvocation(message: CallClientMessage): Promise<void> {
    try {
      const data = await this.registry.invoke(message)

      this.transport.send({
        type: 'callClientResult',
        requestId: message.requestId,
        ok: true,
        data
      })
    } catch (error) {
      const normalized = error instanceof Error ? error : new Error(String(error))

      this.transport.send({
        type: 'callClientResult',
        requestId: message.requestId,
        ok: false,
        error: createSerializedError('handler_error', normalized.message)
      })
    }
  }

  private ensureConnected(): void {
    if (!this.connected) {
      throw new Error('MDP client is not connected')
    }
  }

  private ensureRegistered(): void {
    this.ensureConnected()

    if (!this.registered) {
      throw new Error('MDP client is not registered')
    }
  }

  private async openTransport(): Promise<void> {
    await this.authenticateTransport()
    await this.transport.connect()
    this.connected = true
  }

  private sendRegisterMessage(): void {
    this.transport.send({
      type: 'registerClient',
      client: this.registry.describe(this.clientInfo),
      ...(this.auth ? { auth: this.auth } : {})
    })
  }

  private async handleTransportClose(): Promise<void> {
    const wasConnected = this.connected
    const wasRegistered = this.registered

    this.connected = false
    this.registered = false

    if (!wasConnected) {
      return
    }

    this.restoreRegistrationOnReconnect ||= wasRegistered
    this.reconnectController.handleUnexpectedClose()
  }
}

export function createMdpClient(options: MdpClientOptions): MdpClient {
  return new MdpClient(options)
}

export { resolveServerUrl } from './transport/client-connection.js'

function readLegacyArgs(request: PathRequest) {
  if (request.body !== undefined) {
    if (isArgumentRecord(request.body)) {
      return request.body
    }

    throw new Error('Legacy capability arguments must be a JSON object')
  }

  return request.queries
}

function toLegacyContext(
  request: PathRequest,
  context: PathInvocationContext
): LegacyCapabilityContext {
  return {
    ...context,
    params: request.params,
    queries: request.queries,
    headers: request.headers
  }
}

function isArgumentRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

File: sdks/jvm/java-client/src/main/java/io/modeldriveprotocol/client/MdpClient.java
package io.modeldriveprotocol.client;

import java.net.URI;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.atomic.AtomicBoolean;

public final class MdpClient {
  private volatile ClientInfo clientInfo;
  private volatile AuthContext auth;
  private final ProcedureRegistry registry = new ProcedureRegistry();
  private final ClientTransport transport;
  private final AtomicBoolean connected = new AtomicBoolean(false);
  private final AtomicBoolean registered = new AtomicBoolean(false);

  public MdpClient(String serverUrl, ClientInfo clientInfo) {
    this(serverUrl, clientInfo, createDefaultTransport(serverUrl));
  }

  public MdpClient(String serverUrl, ClientInfo clientInfo, ClientTransport transport) {
    Objects.requireNonNull(serverUrl, "serverUrl");
    this.clientInfo = Objects.requireNonNull(clientInfo, "clientInfo");
    this.transport = Objects.requireNonNull(transport, "transport");
  }

  public MdpClient setAuth(AuthContext auth) {
    this.auth = auth;
    return this;
  }

  public MdpClient exposeEndpoint(
      String path,
      HttpMethod method,
      PathHandler handler,
      EndpointOptions options) {
    registry.exposeEndpoint(path, method, handler, options == null ? new EndpointOptions() : options);
    return this;
  }

  public MdpClient exposeSkillMarkdown(String path, String content, SkillOptions options) {
    registry.exposeSkillMarkdown(path, content, options == null ? new SkillOptions() : options);
    return this;
  }

  public MdpClient exposePromptMarkdown(String path, String content, PromptOptions options) {
    registry.exposePromptMarkdown(path, content, options == null ? new PromptOptions() : options);
    return this;
  }

  public ClientDescriptor describe() {
    return registry.describe(clientInfo);
  }

  public CompletionStage<Void> connect() {
    return transport
        .connect(this::handleMessage, this::handleTransportClose)
        .thenRun(() -> connected.set(true));
  }

  public CompletionStage<Void> register(ClientInfoOverride overrideInfo) {
    requireConnected();
    if (overrideInfo != null) {
      clientInfo = clientInfo.apply(overrideInfo);
    }
    return transport
        .send(ProtocolCodec.buildRegisterClientMessage(describe(), auth))
        .thenRun(() -> registered.set(true));
  }

  public CompletionStage<Void> syncCatalog() {
    requireConnected();
    requireRegistered();
    return transport.send(
        ProtocolCodec.buildUpdateClientCatalogMessage(
            clientInfo.id(),
            registry.describePaths()));
  }

  public CompletionStage<Void> disconnect() {
    CompletionStage<Void> unregister =
        connected.get() && registered.get()
            ? transport.send(ProtocolCodec.buildUnregisterClientMessage(clientInfo.id()))
            : CompletableFuture.completedFuture(null);
    return unregister
        .handle((ignored, error) -> null)
        .thenCompose(ignored -> transport.close())
        .thenRun(
            () -> {
              connected.set(false);
              registered.set(false);
            });
  }

  private void handleMessage(ProtocolCodec.ServerToClientMessage message) {
    if (message instanceof ProtocolCodec.PingMessage pingMessage) {
      transport.send(ProtocolCodec.buildPongMessage(pingMessage.timestamp()));
      return;
    }

    if (message instanceof ProtocolCodec.CallClientMessage callClientMessage) {
      registry
          .invoke(callClientMessage)
          .handle(
              (data, error) ->
                  error == null
                      ? ProtocolCodec.buildCallClientResultMessage(
                          callClientMessage.requestId(),
                          true,
                          data,
                          null)
                      : ProtocolCodec.buildCallClientResultMessage(
                          callClientMessage.requestId(),
                          false,
                          null,
                          SerializedError.handler(error.getCause() == null ? error.getMessage() : error.getCause().getMessage())))
          .thenCompose(transport::send);
    }
  }

  private void handleTransportClose() {
    connected.set(false);
    registered.set(false);
  }

  private void requireConnected() {
    if (!connected.get()) {
      throw new IllegalStateException("MDP client is not connected");
    }
  }

  private void requireRegistered() {
    requireConnected();
    if (!registered.get()) {
      throw new IllegalStateException("MDP client is not registered");
    }
  }

  private static ClientTransport createDefaultTransport(String serverUrl) {
    String scheme = URI.create(serverUrl).getScheme();
    if ("ws".equals(scheme) || "wss".equals(scheme)) {
      return new WebSocketClientTransport(serverUrl);
    }
    if ("http".equals(scheme) || "https".equals(scheme)) {
      return new HttpLoopClientTransport(serverUrl);
    }
    throw new IllegalArgumentException("Unsupported MDP transport protocol: " + scheme);
  }

  public enum HttpMethod {
    GET,
    POST,
    PUT,
    PATCH,
    DELETE
  }

  public interface PathHandler {
    CompletionStage<Object> handle(PathRequest request, PathInvocationContext context);
  }

  public record AuthContext(
      String scheme,
      String token,
      Map<String, String> headers,
      Map<String, Object> metadata) {
    public Map<String, Object> toMap() {
      Map<String, Object> payload = new LinkedHashMap<>();
      if (scheme != null) {
        payload.put("scheme", scheme);
      }
      if (token != null) {
        payload.put("token", token);
      }
      if (headers != null && !headers.isEmpty()) {
        payload.put("headers", headers);
      }
      if (metadata != null && !metadata.isEmpty()) {
        payload.put("metadata", metadata);
      }
      return payload;
    }
  }

  public record ClientInfo(
      String id,
      String name,
      String description,
      String version,
      String platform,
      Map<String, Object> metadata) {
    public ClientInfo(String id, String name) {
      this(id, name, null, null, null, null);
    }

    public ClientInfo apply(ClientInfoOverride overrideInfo) {
      return new ClientInfo(
          overrideInfo.id != null ? overrideInfo.id : id,
          overrideInfo.name != null ? overrideInfo.name : name,
          overrideInfo.description != null ? overrideInfo.description : description,
          overrideInfo.version != null ? overrideInfo.version : version,
          overrideInfo.platform != null ? overrideInfo.platform : platform,
          overrideInfo.metadata != null ? overrideInfo.metadata : metadata);
    }
  }

  public static final class ClientInfoOverride {
    private final String id;
    private final String name;
    private final String description;
    private final String version;
    private final String platform;
    private final Map<String, Object> metadata;

    public ClientInfoOverride() {
      this(null, null, null, null, null, null);
    }

    public ClientInfoOverride(
        String id,
        String name,
        String description,
        String version,
        String platform,
        Map<String, Object> metadata) {
      this.id = id;
      this.name = name;
      this.description = description;
      this.version = version;
      this.platform = platform;
      this.metadata = metadata;
    }

    public ClientInfoOverride description(String value) {
      return new ClientInfoOverride(id, name, value, version, platform, metadata);
    }
  }

  public sealed interface PathDescriptor permits EndpointPathDescriptor, SkillPathDescriptor, PromptPathDescriptor {
    String path();

    String descriptorType();

    Map<String, Object> toMap();
  }

  public record EndpointPathDescriptor(
      String path,
      HttpMethod method,
      String description,
      Map<String, Object> inputSchema,
      Map<String, Object> outputSchema,
      String contentType)
      implements PathDescriptor {
    @Override
    public String descriptorType() {
      return "endpoint";
    }

    @Override
    public Map<String, Object> toMap() {
      Map<String, Object> payload = new LinkedHashMap<>();
      payload.put("type", "endpoint");
      payload.put("path", path);
      payload.put("method", method.name());
      if (description != null) {
        payload.put("description", description);
      }
      if (inputSchema != null) {
        payload.put("inputSchema", inputSchema);
      }
      if (outputSchema != null) {
        payload.put("outputSchema", outputSchema);
      }
      if (contentType != null) {
        payload.put("contentType", contentType);
      }
      return payload;
    }
  }

  public record SkillPathDescriptor(String path, String description, String contentType)
      implements PathDescriptor {
    @Override
    public String descriptorType() {
      return "skill";
    }

    @Override
    public Map<String, Object> toMap() {
      Map<String, Object> payload = new LinkedHashMap<>();
      payload.put("type", "skill");
      payload.put("path", path);
      payload.put("contentType", contentType);
      if (description != null) {
        payload.put("description", description);
      }
      return payload;
    }
  }

  public record PromptPathDescriptor(
      String path,
      String description,
      Map<String, Object> inputSchema,
      Map<String, Object> outputSchema)
      implements PathDescriptor {
    @Override
    public String descriptorType() {
      return "prompt";
    }

    @Override
    public Map<String, Object> toMap() {
      Map<String, Object> payload = new LinkedHashMap<>();
      payload.put("type", "prompt");
      payload.put("path", path);
      if (description != null) {
        payload.put("description", description);
      }
      if (inputSchema != null) {
        payload.put("inputSchema", inputSchema);
      }
      if (outputSchema != null) {
        payload.put("outputSchema", outputSchema);
      }
      return payload;
    }
  }

  public record ClientDescriptor(
      String id,
      String name,
      List<PathDescriptor> paths,
      String description,
      String version,
      String platform,
      Map<String, Object> metadata) {
    public Map<String, Object> toMap() {
      Map<String, Object> payload = new LinkedHashMap<>();
      payload.put("id", id);
      payload.put("name", name);
      payload.put("paths", paths.stream().map(PathDescriptor::toMap).toList());
      if (description != null) {
        payload.put("description", description);
      }
      if (version != null) {
        payload.put("version", version);
      }
      if (platform != null) {
        payload.put("platform", platform);
      }
      if (metadata != null) {
        payload.put("metadata", metadata);
      }
      return payload;
    }
  }

  public record SerializedError(String code, String message, Object details) {
    static SerializedError handler(String message) {
      return new SerializedError("handler_error", message, null);
    }

    Map<String, Object> toMap() {
      Map<String, Object> payload = new LinkedHashMap<>();
      payload.put("code", code);
      payload.put("message", message);
      if (details != null) {
        payload.put("details", details);
      }
      return payload;
    }
  }

  public record PathRequest(
      Map<String, Object> params,
      Map<String, Object> queries,
      Object body,
      Map<String, String> headers) {}

  public record PathInvocationContext(
      String requestId,
      String clientId,
      String pathType,
      HttpMethod method,
      String path,
      AuthContext auth) {}

  public record EndpointOptions(
      String description,
      Map<String, Object> inputSchema,
      Map<String, Object> outputSchema,
      String contentType) {
    public EndpointOptions() {
      this(null, null, null, null);
    }

    public EndpointOptions description(String value) {
      return new EndpointOptions(value, inputSchema, outputSchema, contentType);
    }
  }

  public record SkillOptions(String description, String contentType) {
    public SkillOptions() {
      this(null, "text/markdown");
    }

    public SkillOptions description(String value) {
      return new SkillOptions(value, contentType);
    }
  }

  public record PromptOptions(
      String description,
      Map<String, Object> inputSchema,
      Map<String, Object> outputSchema) {
    public PromptOptions() {
      this(null, null, null);
    }

    public PromptOptions description(String value) {
      return new PromptOptions(value, inputSchema, outputSchema);
    }
  }

  private static final class ProcedureRegistry {
    private final List<ProcedureEntry> entries = new ArrayList<>();

    synchronized void exposeEndpoint(
        String path,
        HttpMethod method,
        PathHandler handler,
        EndpointOptions options) {
      register(
          new EndpointPathDescriptor(
              path,
              method,
              options.description(),
              options.inputSchema(),
              options.outputSchema(),
              options.contentType()),
          handler);
    }

    synchronized void exposeSkillMarkdown(String path, String content, SkillOptions options) {
      String description =
          options.description() != null ? options.description() : deriveMarkdownDescription(content);
      register(
          new SkillPathDescriptor(path, description, options.contentType()),
          (_request, _context) -> CompletableFuture.completedFuture(content));
    }

    synchronized void exposePromptMarkdown(String path, String content, PromptOptions options) {
      String description =
          options.description() != null ? options.description() : deriveMarkdownDescription(content);
      register(
          new PromptPathDescriptor(path, description, options.inputSchema(), options.outputSchema()),
          (_request, _context) ->
              CompletableFuture.completedFuture(
                  Map.of("messages", List.of(Map.of("role", "user", "content", content)))));
    }

    synchronized List<PathDescriptor> describePaths() {
      return entries.stream().map(ProcedureEntry::descriptor).toList();
    }

    synchronized ClientDescriptor describe(ClientInfo clientInfo) {
      return new ClientDescriptor(
          clientInfo.id(),
          clientInfo.name(),
          describePaths(),
          clientInfo.description(),
          clientInfo.version(),
          clientInfo.platform(),
          clientInfo.metadata());
    }

    CompletionStage<Object> invoke(ProtocolCodec.CallClientMessage message) {
      ProcedureEntry entry = resolveEntry(message.method(), message.path());
      if (entry == null) {
        return CompletableFuture.failedFuture(
            new IllegalArgumentException(
                "Unknown path \"" + message.path() + "\" for method \"" + message.method() + "\""));
      }
      PathRequest request =
          new PathRequest(
              entry.match.params(),
              message.query() == null ? Map.of() : message.query(),
              message.body(),
              message.headers() == null ? Map.of() : message.headers());
      PathInvocationContext context =
          new PathInvocationContext(
              message.requestId(),
              message.clientId(),
              entry.descriptor.descriptorType(),
              message.method(),
              message.path(),
              message.auth());
      return entry.handler.handle(request, context);
    }

    private void register(PathDescriptor descriptor, PathHandler handler) {
      assertDescriptorPathShape(descriptor);
      String key = registrationKey(descriptor);
      ProcedureEntry entry = new ProcedureEntry(descriptor, handler, null);
      for (int index = 0; index < entries.size(); index += 1) {
        if (registrationKey(entries.get(index).descriptor).equals(key)) {
          entries.set(index, entry);
          return;
        }
      }
      entries.add(entry);
    }

    private ProcedureEntry resolveEntry(HttpMethod method, String path) {
      ProcedureEntry bestMatch = null;
      for (ProcedureEntry entry : entries) {
        if (!matchesMethod(entry.descriptor, method)) {
          continue;
        }
        PathUtils.PathPatternMatch match = PathUtils.matchPathPattern(entry.descriptor.path(), path);
        if (match == null) {
          continue;
        }
        if (bestMatch == null
            || PathUtils.comparePathSpecificity(match.specificity(), bestMatch.match.specificity()) > 0) {
          bestMatch = new ProcedureEntry(entry.descriptor, entry.handler, match);
        }
      }
      return bestMatch;
    }

    private void assertDescriptorPathShape(PathDescriptor descriptor) {
      if (!PathUtils.isPathPattern(descriptor.path())) {
        throw new IllegalArgumentException("Invalid path: " + descriptor.path());
      }

      if (descriptor instanceof EndpointPathDescriptor endpoint) {
        if (PathUtils.isSkillPath(endpoint.path()) || PathUtils.isPromptPath(endpoint.path())) {
          throw new IllegalArgumentException("Endpoint path cannot target reserved skill or prompt leaves");
        }
      }

      if (descriptor instanceof SkillPathDescriptor skill && !PathUtils.isSkillPath(skill.path())) {
        throw new IllegalArgumentException("Skill paths must end with skill.md or SKILL.md");
      }

      if (descriptor instanceof PromptPathDescriptor prompt && !PathUtils.isPromptPath(prompt.path())) {
        throw new IllegalArgumentException("Prompt paths must end with prompt.md or PROMPT.md");
      }
    }

    private String registrationKey(PathDescriptor descriptor) {
      if (descriptor instanceof EndpointPathDescriptor endpoint) {
        return endpoint.method() + " " + endpoint.path();
      }
      return descriptor.path();
    }

    private boolean matchesMethod(PathDescriptor descriptor, HttpMethod method) {
      if (descriptor instanceof EndpointPathDescriptor endpoint) {
        return endpoint.method() == method;
      }
      return method == HttpMethod.GET;
    }

    private String deriveMarkdownDescription(String content) {
      List<String> paragraph = new ArrayList<>();
      for (String rawLine : content.split("\\R")) {
        String line = rawLine.trim();
        if (line.isEmpty()) {
          if (!paragraph.isEmpty()) {
            break;
          }
          continue;
        }
        if (line.startsWith("#")) {
          continue;
        }
        paragraph.add(line);
      }
      return paragraph.isEmpty() ? null : String.join(" ", paragraph);
    }

    private record ProcedureEntry(
        PathDescriptor descriptor,
        PathHandler handler,
        PathUtils.PathPatternMatch match) {}
  }
}
